### PR TITLE
Framework for multi-tenancy support

### PR DESCRIPTION
**Pull request by codeforall**
The PR aims to build multi-tenancy support and contains multiple commits for the following.

- Establishing a Framework for Master Key.
- New Shared Cache Management Framework.
- Adding catalog table for managing key providers.
- Revamping the Keyring API Interface and Integrating Master Key.

**Individual commit messages contain a more detailed description of changes.**

Few Notes:

- A few test cases might fail with PR, but that is because of the file location of the test file vault and not because of any functionality issue. We can take care of that as a separate commit after discussion.
- Currently, the PR only creates a framework and implements a mechanism to install a key provider and master keys for a database. Altering, Deleting, and handling master-key dependency on key provider will be taken care of as a separate PR
- Unifying master-key-info with keymap file is also a TODO item for the next PR

### CODE REVIEW

File: sql/move_large_tuples.sql [hunk 1]
```diff
@@ -1,6 +1,9 @@
 -- test pg_tde_move_encrypted_data()
 CREATE EXTENSION pg_tde;
 
+SELECT pg_tde_add_key_provider_file('file-valut','/tmp/pg_tde_test_keyring.per');
```
Comment (dutow) [inline sql/move_large_tuples.sql:4]: Here and in other files: `valut` looks like a typo

File: sql/toast_extended_storage.sql [hunk 2]
```diff
@@ -1,6 +1,9 @@
 -- test https://github.com/Percona-Lab/pg_tde/issues/63
 CREATE EXTENSION pg_tde;
 
+SELECT pg_tde_add_key_provider_file('file-valut','/tmp/pg_tde_test_keyring.per');
+--SELECT pg_tde_set_master_key('test-db-master-key','file-valut');
```
Comment (dutow) [inline sql/toast_extended_storage.sql:5]: Why is this commented out?

File: src/catalog/tde_keyring.c [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+	provider_id = DatumGetInt32(datum);
+
+	datum = heap_getattr(tuple, PG_TDE_KEY_PROVIDER_TYPE_ATTRNUM, tupDesc, &isnull);
+	keyring_type = TextDatumGetCString(datum);
+
+	datum = heap_getattr(tuple, PG_TDE_KEY_PROVIDER_NAME_ATTRNUM, tupDesc, &isnull);
+	keyring_name = TextDatumGetCString(datum);
+
+	option_datum = heap_getattr(tuple, PG_TDE_KEY_PROVIDER_OPTIONS_ATTRNUM, tupDesc, &isnull);
+	/*keyring_options = TextDatumGetCString(option_datum);*/
```
Comment (dutow) [inline src/catalog/tde_keyring.c:90]: Why is this commented out?

File: src/catalog/tde_keyring.c [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+static FileKeyring *
+load_file_keyring_provider_options(Datum keyring_options)
+{
+
+	Datum file_path;
+	FileKeyring *file_keyring = palloc0(sizeof(FileKeyring));
+	/*
+	Datum file_type;
+	file_type = DirectFunctionCall2(json_object_field_text, keyring_options, CStringGetTextDatum(FILE_KEYRING_TYPE_KEY));
```
Comment (dutow) [inline src/catalog/tde_keyring.c:226]: Why the duplicated commented out code?

File: src/catalog/tde_keyring.c [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+	*/
+	file_path = DirectFunctionCall2(json_object_field_text, keyring_options, CStringGetTextDatum(FILE_KEYRING_PATH_KEY));
+	/* TODO check NULL  and verify type */
+	file_keyring->keyring.type = FILE_KEY_PROVIDER;
+	strncpy(file_keyring->file_name, TextDatumGetCString(file_path), sizeof(file_keyring->file_name));
+	return file_keyring;
+}
+
+static ValutV2Keyring *
+load_valutV2_keyring_provider_options(Datum keyring_options)
```
Comment (dutow) [inline src/catalog/tde_keyring.c:236]: typo: `valut`

File: src/catalog/tde_master_key.c [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+    if (keyInfo == NULL)
+    {
+        ereport(ERROR,
+                (errmsg("failed to retrieve master key from keyring")));
+        return NULL;
+    }
+
+    masterKey = palloc(sizeof(TDEMasterKey));
+    masterKey->databaseId = MyDatabaseId;
+    masterKey->keyVersion = 1;
```
Comment (dutow) [inline src/catalog/tde_master_key.c:324]: Shouldn't we read these values from the file?

File: src/catalog/tde_master_key.c [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+    TDEMasterKeyInfo *masterKeyInfo = NULL;
+    File master_key_file = -1;
+    off_t bytes_written = 0;
+    char *info_file_path = get_master_key_info_path();
+
+    Assert(master_key != NULL);
+    Assert(keyring != NULL);
+
+    masterKeyInfo = palloc(sizeof(TDEMasterKeyInfo));
+    masterKeyInfo->keyId = 0;
```
Comment (dutow) [inline src/catalog/tde_master_key.c:205]: Why these hardcoded values here? `master_key` already contains these fields, shoulndt we use the values from it?

File: src/include/catalog/tde_keyring.h [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+	FILE_KEY_PROVIDER,
+	VAULT_V2_KEY_PROVIDER,
+} ProviderType;
+
+/* Base type for all keyring */
+typedef struct GenericKeyring
+{
+	ProviderType type; /* Must be the first field */
+	Oid keyId;
+	char keyName[128];
```
Comment (dutow) [inline src/include/catalog/tde_keyring.h:37]: Shouldn't we add a define for 128?

File: src/include/catalog/tde_master_key.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+#include "keyring/keyring_api.h"
+#include "nodes/pg_list.h"
+
+#define MASTER_KEY_NAME_LEN TDE_KEY_NAME_LEN
+
+typedef struct TDEMasterKey
+{
+	Oid databaseId;
+	uint32 keyVersion;
+    Oid keyringId;
```
Comment (dutow) [inline src/include/catalog/tde_master_key.h:25]: indentation

File: src/include/keyring/keyring_api.h [hunk 10]
```diff
@@ -1,16 +1,28 @@
...
 typedef struct keyName
 {
-	char name[256]; // enough for now
+	char name[TDE_KEY_NAME_LEN];
 } keyName;
 
 typedef struct keyData
 {
-	unsigned char data[32]; // maximum 256 bit encryption
+	unsigned char data[MAX_KEY_DATA_SIZE]; /* maximum 256 bit encryption */
```
Comment (dutow) [inline src/include/keyring/keyring_api.h:25]: I don't think we need this comment at both places

File: src/include/keyring/keyring_api.h [hunk 11]
```diff
@@ -20,6 +32,29 @@ typedef struct keyInfo
...
+	keyInfo *(*keyring_get_key)(GenericKeyring *keyring, const char *key_name, bool throw_error, KeyringReturnCodes *returnCode);
+	KeyringReturnCodes (*keyring_store_key)(GenericKeyring *keyring, keyInfo *key, bool throw_error);
+} TDEKeyringRoutine;
+
+extern bool RegisterKeyProvider(const TDEKeyringRoutine *routine, ProviderType type);
+
+extern KeyringReturnCodes KeyringStoreKey(GenericKeyring *keyring, keyInfo *key, bool throw_error);
+extern keyInfo *KeyringGetKey(GenericKeyring *keyring, const char *key_name, bool throw_error, KeyringReturnCodes *returnCode);
+
+extern keyInfo *keyringGenerateNewKeyAndStore(GenericKeyring *keyring, const char *key_name, unsigned key_len, bool throw_error);
```
Comment (dutow) [inline src/include/keyring/keyring_api.h:55]: `KeyringGenerate` to match the other functions?

File: src/keyring/keyring_file.c [hunk 12]
```diff
@@ -1,75 +1,114 @@
...
+
+    file = PathNameOpenFile(file_keyring->file_name, O_CREAT | O_EXCL | O_RDWR | PG_BINARY);
+    if (file < 0)
+    {
+		ereport(throw_error?ERROR:WARNING,
+			(errcode_for_file_access(),
+			errmsg("Failed to open keyring file %s :%m", file_keyring->file_name)));
+        return KEYRING_CODE_RESOURCE_NOT_ACCESSABLE;
+    }
+	/* Write key to the end of file */
```
Comment (dutow) [inline src/keyring/keyring_file.c:100]: Shouldn't this function also report an error if the key already exists in the file?

File: pg_tde--1.0.sql [hunk 13]
```diff
@@ -3,6 +3,62 @@
 -- complain if script is sourced in psql, rather than via CREATE EXTENSION
 \echo Use "CREATE EXTENSION pg_tde" to load this file. \quit
 
+-- pg_tde catalog tables
+CREATE SCHEMA percona_tde;
+-- Note: The table is created using heap storage becasue we do not want this table
+-- to be encrypted by pg_tde. This table is used to store key provider information
+-- and we do not want to encrypt this table using pg_tde.
```
[the remark this answers is not in the document]
Comment (codeforall) [reply pg_tde--1.0.sql:10]: Yes that part needs some thoughts and discussions. Will take care of that after this PR

File: sql/move_large_tuples.sql (hunk 1 above)
Comment (codeforall) [reply sql/move_large_tuples.sql]: Fixed!

File: sql/toast_extended_storage.sql (hunk 2 above)
Comment (codeforall) [reply sql/toast_extended_storage.sql]: Right now we need to figure out a way to delete the files created by extension if the extension is deleted. So I have intentionally commented out these lines to take care of that. IMHO we can handle this as a separate PR

File: src/catalog/tde_keyring.c (hunk 3 above)
Comment (codeforall) [reply src/catalog/tde_keyring.c]: Fixed!.

File: src/include/catalog/tde_keyring.h (hunk 8 above)
Comment (codeforall) [reply src/include/catalog/tde_keyring.h]: Done!

File: src/include/catalog/tde_master_key.h (hunk 9 above)
Comment (codeforall) [reply src/include/catalog/tde_master_key.h]: Done!

File: src/include/keyring/keyring_api.h (hunk 10 above)
Comment (codeforall) [reply src/include/keyring/keyring_api.h]: Done!

File: src/keyring/keyring_file.c (hunk 12 above)
Comment (codeforall) [reply src/keyring/keyring_file.c:112]: You are right! Committed the fix

File: src/catalog/tde_master_key.c [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+    if (acquire_lock)
+        LWLockRelease(shared_state->Lock);
+    return cacheEntry;
+}
+
+/*
+ * push the master key for current database to the shared memory cache
+ */
+static void
+push_master_key_to_cache(TDEMasterKey *masterKey)
```
Comment (dAdAbird) [inline src/catalog/tde_master_key.c:481]: As a TODO for the future: set a limit on the cache size (and add eviction policy). As I understand, now cache might grow indefinitely

**Review by dAdAbird — COMMENTED** [review]
It also makes to run `pgindent` on new files.

File: src/catalog/tde_master_key.c (hunk 14 above)
Comment (codeforall) [reply src/catalog/tde_master_key.c:488]: That is a correct observation. At the same time, we would only have one cache entry for each database, so I am not expecting the cache to grow insanely huge, So IMHO we can take care of this at a later stage. For now I have added a TODO note on the function.

File: pg_tde--1.0.sql [hunk 15]
```diff
@@ -18,9 +74,12 @@ RETURNS boolean
 AS 'MODULE_PATHNAME'
 LANGUAGE C;
 
+CREATE FUNCTION pg_tde_set_master_key(master_key_name VARCHAR(255), provider_name VARCHAR(255))
```
[the remark this answers is not in the document]
Comment (codeforall) [reply pg_tde--1.0.sql:77]: Thanks for pointing it out. I've updated the catalog definition.

File: src/include/keyring/keyring_api.h (hunk 11 above)
Comment (dutow) [reply src/include/keyring/keyring_api.h:55]: Not fixed, the `k` is still small, this file now has functions with different naming conventions

File: sql/multi_insert.sql [hunk 16]
```diff
@@ -2,6 +2,9 @@
 -- 
 CREATE EXTENSION pg_tde;
 
+SELECT pg_tde_add_key_provider_file('file-vault','/tmp/pg_tde_test_keyring.per');
```
Comment (dutow) [inline sql/multi_insert.sql:5]: I realized an issue with these changes: now the tests are hardcoded to only run with keyring file, previously we ran them both with keyring file and vault. With these changes, effectively we no longer execute vault tests.

If we go in this direction, we have to duplicate all test files, and later when we add additional providers, copy and copy them again and again. I don't think this is a good approach, but I do not see how we can do it differently, as the tests would require SQL commands before execution.

**Review by dutow — COMMENTED** [review]
I see no changes for keyring vault. Will it continue to work after these changes?

File: src/keyring/keyring_file.c [hunk 17]
```diff
@@ -1,75 +1,126 @@
...
 
-	if(!json_object_object_get_ex(configRoot, "datafile", &dataO))
-	{
-		elog(ERROR, "Missing 'datafile' attribute.");
-		return 0;
-	}
 
-	datafile = keyringParseStringParam(dataO);
+static keyInfo*
+get_key_by_name(GenericKeyring* keyring, const char* key_name, bool throw_error, KeyringReturnCodes *returnCode)
```
[the remark this answers is not in the document]
Comment (codeforall) [reply src/keyring/keyring_file.c]: Fixed

File: src/catalog/tde_keyring.c [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+	keyring_name = TextDatumGetCString(datum);
+
+	option_datum = heap_getattr(tuple, PG_TDE_KEY_PROVIDER_OPTIONS_ATTRNUM, tupDesc, &isnull);
+
+	provider_type = get_keyring_provider_from_typename(keyring_type);
+
+	keyring = load_keyring_provider_options(provider_type, option_datum);
+	if (keyring)
+	{
+		strncpy(keyring->keyName, keyring_name, sizeof(keyring->keyName));
```
[the remark this answers is not in the document]
Comment (codeforall) [reply src/catalog/tde_keyring.c]: Catalog definition ensures that we don't get NULL keyring name

File: src/catalog/tde_keyring.c [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+static FileKeyring *
+load_file_keyring_provider_options(Datum keyring_options)
+{
+
+	Datum file_path;
+	FileKeyring *file_keyring = palloc0(sizeof(FileKeyring));
+	file_path = DirectFunctionCall2(json_object_field_text, keyring_options, CStringGetTextDatum(FILE_KEYRING_PATH_KEY));
+	/* TODO check NULL  and verify type */
```
[the remark this answers is not in the document]
Comment (codeforall) [reply src/catalog/tde_keyring.c]: Adjusted the function definition to ensure we do not get a NULL JSON value

**Review by dutow — APPROVED** [review]

**Pull request merged**